src/scalems/radical/__init__.py

```diff
@@ -1204,44 +1232,44 @@ def runtime_startup(self, runner_started: asyncio.Event) -> asyncio.Task:
             scalems.execution.manage_execution(
                 self,
                 processing_state=runner_started))
+        # TODO: Note the expected scope of the runner_task lifetime with respect to
+        #  the global state changes (i.e. ContextVars and locks).
         return runner_task
 
-    def runtime_shutdown(self, session: rp.Session):
+    def runtime_shutdown(self, runtime: Runtime):
+        session = getattr(runtime, 'session', None)
         if session is None or session.closed:
             logger.error('Runtime Session is already closed?!')
-        # Cancel the master.
-        logger.debug('Canceling the master scheduling task.')
-        task_manager = session.get_task_managers(tmgr_uids=self._task_manager_uid)
-        if self.scheduler is not None:
-            task_manager.cancel_tasks(uids=self.scheduler.uid)
-        # Cancel blocks until the task is done so the following wait would (currently)
-        # be redundant,
-        # but there is a ticket open to change this behavior.
-        # See https://github.com/radical-cybertools/radical.pilot/issues/2336
-        # tmgr.wait_tasks([scheduler.uid])
-        logger.debug('Master scheduling task complete.')
-
-        pmgr: rp.PilotManager = session.get_pilot_managers(
-            pmgr_uids=self._pilot_manager_uid)
-        # TODO: We may have multiple pilots.
-        # TODO: Check for errors?
-        logger.debug('Canceling Pilot.')
-        pmgr.cancel_pilots(uids=self.pilot_uid)
-        logger.debug('Pilot canceled.')
+        else:
+            # Cancel the master.
+            logger.debug('Canceling the master scheduling task.')
+            task_manager = runtime.task_manager()
```

Comment:
Here, for instance, RP does not guarantee that the TaskManager can be retrieved by UID if the client has not retained a reference. (What would be the result of `Session.get_task_managers()` in such a failure mode?)

@andre-merzky confirms that `scalems.radical.Runtime` should hold actual object references to the managers, and we should probably encapsulate these startup and shutdown logic into the `Runtime` class.

Reply:
My apologies - I forgot that we moved `get_xxx_managers` into the public API.  As such, we *do* guarantee that the objects are alive even if the application will not keep any references.

Reply:
Do you think that https://github.com/SCALE-MS/scale-ms/commit/501c00a80266ac9e26c87d11bee071e7756454f4 is misguided?

Reply:
> Do you think that [501c00a](https://github.com/SCALE-MS/scale-ms/commit/501c00a80266ac9e26c87d11bee071e7756454f4) is misguided?

I suppose it is a slight simplification, and therefore possibly worthwhile on that note alone. It may simply be reasonable, and a sensible precursor to putting the responsibility of `pilot.cancel()`, `pilot_manager.close()`, and `task_manager.close()` into `scalems.radical.Runtime`.

Reply:
That commit looks reasonable to me, would have done the same.